src/python/pants/engine/scheduler.py

```diff
@@ -297,16 +297,32 @@ def with_fork_context(self, func):
 
   def _run_and_return_roots(self, session, execution_request):
     raw_roots = self._native.lib.scheduler_execute(self._scheduler, session, execution_request)
+    remaining_runtime_exceptions_to_capture = list(self._native.consume_cffi_extern_method_runtime_exceptions())
     try:
       roots = []
       for raw_root in self._native.unpack(raw_roots.nodes_ptr, raw_roots.nodes_len):
+        # Check if there were any uncaught exceptions within rules that were executed.
+        remaining_runtime_exceptions_to_capture.extend(self._native.consume_cffi_extern_method_runtime_exceptions())
+
         if raw_root.is_throw:
           state = Throw(self._from_value(raw_root.handle))
+        elif raw_root.handle == self._native.ffi.NULL:
+          # NB: We expect all NULL handles to correspond to uncaught exceptions which are collected
```

Comment:
IIUC this just means we're not robust on the rust side yet wrt signal handling (and maybe other unexpected errors). We should always get a `Result`. If that characterization is right, this is just a blanket backstop. If that is right, it may be prudent to continue to have one, but also we should call the shot, file an issue against corruptions like this and TODO it here.

Reply:
Sorry for the late response -- I believe this behavior is actually the fault of `cffi`, where the `onerror` method will be called to handle exceptions raised by signal handlers defined in python (see https://cffi.readthedocs.io/en/latest/using.html#extern-python-reference), and a null handle seems appropriate enough to return since `cffi` assumes that the `onerror` callback (which we set in `native.py`) has handled the exception.

I believe this PR actually resolves a dissonance we saw in signal handling, where a <kbd>Ctrl-c</kbd> would be handled correctly if it interrupted python code, while it would have caused this nondeterministic null handle error if it interrupted rust code. @blorente may have thoughts on whether that would make sense.